### PR TITLE
Raise error when training on variable-horizon tasks in AIRL, GAIL and DRLHP

### LINKED ISSUE #324: GAIL/AIRL: fix bias discussed in discriminator-actor-critic paper
This is essentially the same issue as #40 but I'm creating a new one rather than reopening because the code base has changed significantly and some of the discussion there is not very relevant anymore.

The [discriminator-actor-critic paper](https://arxiv.org/abs/1809.02925) describes a bias that GAIL (and apparently to a lesser extent also AIRL) suffer from in a variable-length horizon setting: in GAIL, the learned reward can never be negative, which biases the agent to aim for longer episodes. This is undesirable in some environments, where some goal should be achieved quickly. The DAC paper suggests fixing this bias by learning a reward for the final state rather than fixing it to zero.

#40 was closed under the assumption that #184 fixes the bias but as far as I can tell, that's not the case:
- GAIL doesn't use `RewardNet` at all, so it definitely still has this bias. What we need to do here is to add the fix inside the discriminator network, where the reward is computed
- For AIRL the issue is not the reward model, which can learn negative rewards anyway. If I understand the argument in the DAC paper correctly, this is more closely related to the discriminator loss and we should also add a fix inside the discriminator

I think the fix for GAIL is pretty straightforward, we need to learn a final reward and add it in `DiscrimNetGAIL.reward_train()` if `done` is `True`. Could be optional but should arguably be turned on by default? For AIRL I'm less sure what precisely we need to do. But presumably, the bias is also less critical there because the reward model can learn negative rewards.

### PR DESCRIPTION
**Pull request by AdamGleave**
Variable horizon tasks cause all sorts of problems as discussed in #324 and elsewhere. This PR raises `ValueError` if it detects a variable horizon environment while training `AdversarialTrainer` or `PreferenceComparison`. I've not applied this to `BC` or `DAgger` as I think they don't suffer from this issue, since they're trying to mimic transitions. Variable vs fixed horizon will change the distribution over transitions a bit, which may change training dynamics, but I don't think in a systematic way (limit of infinite data, minimal loss model should still be the same). I also haven't done anything for `mce_irl` as I think it's fixed horizon by design (our tabular environments specify a horizon).

I've also added a rant to the documentation about why variable horizon environments are bad. Maybe this should be a blog post rather than part of the doc, or I should just tone it down? Feedback on this welcomed.

I implemented the check via introducing a new abstract base class `BaseImitationAlgorithm` with the check as a helper method. Main reason for this is the check being stateful lets us check trajectory length are the same across multiple calls to the function. I could implement this as a mix-in instead, but I lean towards thinking this is simpler, given most imitation algorithms want this check. I also suspect we may want `BaseImitationAlgorithm` to do more over time as we try to unify the interface (currently the only other thing it does is hold the logger).

One tricky part of this PR was that `AdversarialTraining` and `PreferenceComparison` both by default used full trajectories plus any partial trajectories we gathered during RL training at the point of training the reward model. This means if we do the naive thing of taking all these trajectories and checking they're the same length they, well, won't be. I fixed this by applying the check only to full trajectories in `AdversarialTraining`, and in `PreferenceComparison` only returning full trajectories.

I'm not totally happy with this solution though. It's wasteful to discard partial trajectories. But over-sampling early transitions in a trajectory is also weird. I'm not sure which is better. It's probably not right that `AdversarialTraining` and `PreferenceComparison` now have a different default behavior here, though. Another area I'd appreciate feedback.

Fixes #324.

### CODE REVIEW

**Review by ejnnr — COMMENTED** [review]
Thanks for taking care of this! And +1 re using a base class rather than a mixin for `imitation` algorithms, I agree that this might make it easier to unify more things in the future.

General comments in addition to the inline ones:
- I think several examples/scripts will now output the scary new warning by default, they should probably be changed to use `seals_cartpole`. From a quick search, I think at least `README.md`, `gettingstarted.rst` (in the docs) and `examples/quickstart.sh` should be adapted. Experiment scripts are probably not that important. The more radical alternative would be to make the `cartpole` `named_config` actually use the seals version and then have something like `original_cartpole`. But I think that's too extreme, we want users to get fixed horizons by default but they should at least know about it. We should probably add a single sentence about what this "seals" environment is in the README (can link to the new doc page for the reasoning of why this makes sense)
- The tone of the new documentation page seems fine to me. If anything, I'd consider renaming `variable_horizon_footgun` to `allow_variable_horizon` though I think either one is ok
- Having a blog post would be nice, especially if the goal is to make others more aware of this issue - I'd assume a blog post would be more widely read and shared even by people who don't use `imitation`. But even if we had a blog post, I'd also keep around something like the current summary of the issue in the documentation. A blog post would presumably be a bit longer and I think the documentation should link there but still be mostly self-contained.
- I agree it would be nice to have a common convention re using or discarding partial trajectories. Don't have a clear opinion on the bias/wastefulness trade-off. One weak argument for discarding trajectories would be that this is less messy to implement for `PreferenceComparisons` (at least excluding them from the check seems a bit messy after thinking about it for 1 minute). Another one is that `generate_trajectories` always returns complete trajectories, so this would be more consistent (especially for `PreferenceComparisons`, where some trajectories might come from training and some from `generate_trajectories`)

File: src/imitation/algorithms/base.py [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+
+    def __getstate__(self):
+        state = self.__dict__.copy()
+        # logger can't be pickled as it depends on open files
+        del state["_logger"]
+        return state
+
+    def __setstate__(self, state):
+        self.__dict__.update(state)
+        # callee should call set_logger if they want to override this
```
[the remark this answers is not in the document]
Comment (AdamGleave) [reply src/imitation/algorithms/base.py:102]: true, fixed

File: src/imitation/algorithms/preference_comparisons.py [hunk 2]
```diff
@@ -238,7 +239,7 @@ class PreferenceDataset(th.utils.data.Dataset):
     and a probability that fragment 1 is preferred over fragment 2.
 
     This dataset is meant to be generated piece by piece during the
-    training process, which is why data can be added via the .push()
+    training process, which is why testdata can be added via the .push()
```
[the remark this answers is not in the document]
Comment (AdamGleave) [reply src/imitation/algorithms/preference_comparisons.py:242]: thanks

File: src/imitation/scripts/config/train_adversarial.py [hunk 3]
```diff
@@ -125,6 +125,7 @@ def acrobot():
 def cartpole():
     env_name = "CartPole-v1"
     rollout_hint = "cartpole"
+    algorithm_kwargs = {"shared": {"variable_horizon_footgun": True}}
```
[the remark this answers is not in the document]
Comment (AdamGleave) [reply src/imitation/scripts/config/train_adversarial.py:128]: Good catch. I've updated others to use `seals` versions, or add the flag if no `seals`-equivalent exists.

**Comment by AdamGleave** [comment]
Thanks for the review, I think I've addressed the comments raised, let me know if you spot anything new!

**Review by ejnnr — APPROVED** [review]
Thanks for the changes! LGTM (except for the things I already mentioned in #335, but I assume you'll merge that first anyway)

**Pull request merged**